eng/common/scripts/Helpers/Sparse-Clone-Helpers.ps1

```diff
@@ -0,0 +1,34 @@
+$sparseCheckoutFile = ".git/info/sparse-checkout"
+
+function GetProjectRootPath([string]$projectDirectory) {
+    Push-Location $projectDirectory
+    try {
+        return git rev-parse --show-toplevel
+    }
+    finally {
+        Pop-Location
+    }
+}
+
+function GetSparseCloneDir([string]$projectDirectory, [string]$projectName, [string]$repoName) {
+    $root = GetProjectRootPath $projectDirectory
+
+    $sparseSpecCloneDir = "$root/../sparse-spec/$repoName/$projectName"
+    New-Item $sparseSpecCloneDir -Type Directory -Force | Out-Null
+    $createResult = Resolve-Path $sparseSpecCloneDir
+    return $createResult
+}
+
+function InitializeSparseGitClone([string]$repo) {
+    git clone --no-checkout --filter=tree:0 $repo .
+    if ($LASTEXITCODE) { exit $LASTEXITCODE }
+    git sparse-checkout init
+    if ($LASTEXITCODE) { exit $LASTEXITCODE }
+    Remove-Item $sparseCheckoutFile -Force
+}
+
+function AddSparseCheckoutPath([string]$subDirectory) {
+    if (!(Test-Path $sparseCheckoutFile) -or !((Get-Content $sparseCheckoutFile).Contains($subDirectory))) {
+        Write-Output $subDirectory >> .git/info/sparse-checkout
+    }

```

Comment:
I would let git handle all this for you.

```suggestion
    git sparse-checkout add $subDirectory
    if ($LASTEXITCODE) { exit $LASTEXITCODE }
```

For the sparse checkout file exists check, it would be better to try to add the path anyway, because you would want the script to fail quickly if you were trying to add a path before initializing the sparse checkout. Otherwise you might think a behavior is happening when it's not.

For the sparseCheckoutFile contains check, the ordering of the sparse checkout file matters. So you could have a path specified for inclusion on line 1, but then another path pattern on line 2 that excludes the path on line 1. So in this case, if a user of this script was building a set of patterns, you wouldn't want to make ordering decisions for them. I think this is important because if we're writing a centralized helper script we'll probably start using it in more places.

Reply:
> if you were trying to add a path before initializing the sparse checkout. 

I'm afraid we are always doing this. After running "git sparse-checkout init", the autogenerated sparse-checkout file contains `/*` and `!/*/`, which are not what we want. So we remove the sparse-checkout file in function `InitializeSparseGitClone`.

> For the sparseCheckoutFile contains check, the ordering of the sparse checkout file matters.

I'm not sure I follows your concern. We don't add any excludes pattern. We don't even add pattern. We only add literal paths.